python/ray/experimental/channel/torch_tensor_nccl_channel.py

```diff
@@ -138,63 +178,130 @@ def write(self, value: Any):
             raise TypeError(msg) from e
         finally:
             # Pop the tensors that were found during serialization of `value`.
-            tensors_to_send = self.serialization_ctx.reset_tensors([])
+            tensors_to_send, _ = self.serialization_ctx.reset_out_of_band_tensors([])
             # Reset the serialization method to now serialize torch.Tensors
             # normally.
             self.serialization_ctx.set_use_external_transport(False)
 
-        # Send the extracted tensors through a GPU-specific channel.
-        self._gpu_data_channel.write(tensors_to_send)
-        # Send the rest of the data, with placeholders for the extracted
-        # tensors, through a CPU-specific channel.
-        self._cpu_data_channel.write(serialized_cpu_data)
+        if self._num_serialized_tensors is not None:
+            # We should only know the number of tensors to expect if the user
+            # specified that the non-tensor data will not change.
+            assert self._static_non_tensor_data
+
+            if self._num_serialized_tensors != len(tensors_to_send):
+                # The serialized value contains a different number of tenors
+                # compared to the previous value that we sent. Throw an error
+                # here instead of writing. Otherwise, the reader would
+                # deserialize a different value from the one we were asked to
+                # write.
+                # NOTE(swang): Unfortunately cloudpickle.dumps is not
+                # deterministic, so directly checking equality of the
+                # serialized values would lead to too many false positives.
+                raise ValueError(
+                    "static_non_tensor_data=True specified, but "
+                    f"the value to write ({value}) contains a different "
+                    "number of tensors"
+                )
+
+        # First send the extracted tensors through a GPU-specific channel.
+        self._tensor_data_channel.write(tensors_to_send)
+
+        # Next send the non-tensor data, if needed. The data contains
+        # placeholders for the extracted tensors, through a CPU-specific
+        # channel.
+        if self._num_serialized_tensors is None:
+            self._non_tensor_data_channel.write(serialized_non_tensor_data)
```

Comment:
Is it guaranteed that this write will be visible to the other device after the prior write (`self._tensor_data_channel.write(tensors_to_send)`) is visible? How is this synchronized?

Reply:
They can be received in either order. Is that what you're asking?